File: legacy/typescript/src/cli/install/CommunityAgentInstallHandler.ts
import {mkdir} from 'node:fs/promises'
import path from 'node:path'

import type {CommunityAgentDefaultConfigBuilder} from '../../config/builder/CommunityAgentDefaultConfigBuilder.js'
import type {CommunityAgentConfigWriter} from '../../config/writer/CommunityAgentConfigWriter.js'
import type {CommunityAgentPathResolver} from '../../config/path/CommunityAgentPathResolver.js'

export interface CommunityInstallResult { readonly configFile: string; readonly inviteUrl: string; readonly botTokenEnvironmentVariable: string }

export class CommunityAgentInstallHandler {
  public constructor(
    private readonly builder: CommunityAgentDefaultConfigBuilder,
    private readonly writer: CommunityAgentConfigWriter,
    private readonly paths: CommunityAgentPathResolver,
  ) {}

  public async install(input: {guildId: string; applicationId: string; force?: boolean}): Promise<CommunityInstallResult> {
    const configFile = this.paths.configFile()
    if (!input.force) {
      try { await import('node:fs/promises').then(fs => fs.access(configFile)); throw new Error(`Config already exists at ${configFile}; use --force to replace it.`) } catch (error) {
        if (error instanceof Error && !error.message.includes('ENOENT')) throw error
      }
    }
    const config = this.builder.build({guildId: input.guildId, applicationId: input.applicationId, dataDirectory: this.paths.dataDirectory()})
    await mkdir(path.dirname(configFile), {recursive: true, mode: 0o700})
    await this.writer.write(configFile, config)
    const permissions = BigInt(1 | 1024 | 2048 | 32768 | 65536 | 131072 | 268435456)
    const inviteUrl = `https://discord.com/oauth2/authorize?client_id=${encodeURIComponent(input.applicationId)}&scope=bot%20applications.commands&permissions=${permissions.toString()}`
    return {configFile, inviteUrl, botTokenEnvironmentVariable: config.discordBotTokenEnvironmentVariable}
  }
}

File: src/main/java/org/tavall/community/install/CommunityAgentInstallHandler.java
package org.tavall.community.install;

import org.tavall.community.config.CommunityAgentConfiguration;
import org.tavall.community.config.CommunityAgentConfigurationWriter;
import org.tavall.community.config.CommunityAgentDefaultConfigurationBuilder;
import org.tavall.community.config.CommunityAgentPathResolver;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;

/** Creates the secure machine configuration and Discord bot invite for a fresh install. */
public final class CommunityAgentInstallHandler {
    private static final long DISCORD_PERMISSIONS = 268_667_905L;

    private final CommunityAgentDefaultConfigurationBuilder configurationBuilder;
    private final CommunityAgentConfigurationWriter configurationWriter;
    private final CommunityAgentPathResolver paths;

    public CommunityAgentInstallHandler(
            CommunityAgentDefaultConfigurationBuilder configurationBuilder,
            CommunityAgentConfigurationWriter configurationWriter,
            CommunityAgentPathResolver paths
    ) {
        this.configurationBuilder = Objects.requireNonNull(configurationBuilder, "configurationBuilder");
        this.configurationWriter = Objects.requireNonNull(configurationWriter, "configurationWriter");
        this.paths = Objects.requireNonNull(paths, "paths");
    }

    public CommunityInstallResult install(
            String guildId,
            String applicationId,
            boolean force
    ) {
        Path configFile = paths.configFile();
        if (Files.exists(configFile) && !force) {
            throw new IllegalStateException(
                    "Config already exists at " + configFile + "; use --force to replace it."
            );
        }

        CommunityAgentConfiguration configuration = configurationBuilder.build(
                guildId,
                applicationId,
                paths.dataDirectory()
        );
        configurationWriter.write(configFile, configuration);
        String inviteUrl = "https://discord.com/oauth2/authorize?client_id="
                + URLEncoder.encode(applicationId, StandardCharsets.UTF_8)
                + "&scope=bot%20applications.commands&permissions="
                + DISCORD_PERMISSIONS;
        return new CommunityInstallResult(
                configFile,
                inviteUrl,
                configuration.discordBotTokenEnvironmentVariable()
        );
    }
}
